sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs

```diff
@@ -0,0 +1,323 @@
+        /// <param name="encryptionKeyWrapAlgorithm">
+        /// Required for uploads. String identifier of the key wrapping algorithm the client should use with
+        /// <paramref name="keyEncryptionKey"/>.
+        /// </param>
+        /// <param name="options">
+        /// Optional client options that define the transport pipeline
+        /// policies for authentication, retries, etc., that are applied to
+        /// every request.
+        /// </param>
+        public EncryptedBlobClient(
+            Uri blobUri,
+            TokenCredential credential,
+            IKeyEncryptionKey keyEncryptionKey = default,
+            IKeyEncryptionKeyResolver keyResolver = default,
+            string encryptionKeyWrapAlgorithm = default,
+            BlobClientOptions options = default)
+            : base(
+                  blobUri,
+                  credential,
+                  options.WithPolicy(new ClientSideDecryptionPolicy(keyResolver, keyEncryptionKey)))
+        {
+            this.KeyWrapper = keyEncryptionKey;
+            this.KeyWrapAlgorithm = encryptionKeyWrapAlgorithm;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BlockBlobClient"/>
+        /// class.
+        /// </summary>
+        /// <param name="blobUri">
+        /// A <see cref="Uri"/> referencing the encrypted block blob that includes the
+        /// name of the account, the name of the container, and the name of
+        /// the blob.
+        /// </param>
+        /// <param name="pipeline">
+        /// The transport pipeline used to send every request.
+        /// </param>
+        internal EncryptedBlobClient(Uri blobUri, HttpPipeline pipeline)
+            : base(blobUri, pipeline)
+        {
+            //TODO make new pipeline for this one
+        }
+        #endregion ctors
+
+        /// <summary>
+        /// This method performs a transform on the content stream for uploads. It is a no-op by default.
+        /// </summary>
+        /// <param name="content">Content to transform.</param>
+        /// <param name="metadata">Content metadata to transform.</param>
+        /// <returns>Transformed content stream.</returns>
+        internal protected override (Stream, Metadata) TransformContent(Stream content, Metadata metadata)
+        {
+            metadata ??= new Dictionary<string, string>();
+
+            var (encryptionStream, encryptionData) = EncryptStream(content);
+
+            using (var stream = new MemoryStream())
+            {
+                var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
+                serializer.WriteObject(stream, encryptionData);
+                stream.Position = 0;
+                metadata.Add(EncryptionConstants.ENCRYPTION_DATA_KEY, new StreamReader(stream).ReadToEnd());
+            }
+
+            return (encryptionStream, metadata);
+        }
+
+        private (Stream, EncryptionData) EncryptStream(Stream plaintext)
```

Comment:
Nit - Would we want a 16th `Stream` class called `StorageEncryptionStream` that carries this data with it?

Reply:
I'm a bit unenthused about that. It makes sense in the case of encryption, but not necessarily in the case of the generic TransformContent method that we override to inject this behavior.